Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h

```diff
@@ -633,131 +635,206 @@ clip(TriangleMesh& tm,
+  // config flags
+  bool clip_volume =
+    parameters::choose_parameter(parameters::get_parameter(np, internal_np::clip_volume), false);
+  bool use_compact_clipper =
+    choose_parameter(get_parameter(np, internal_np::use_compact_clipper), true);
+  const bool throw_on_self_intersection =
+    choose_parameter(get_parameter(np, internal_np::throw_on_self_intersection), false);
+  const bool allow_self_intersections =
+    choose_parameter(get_parameter(np, internal_np::allow_self_intersections), false);
+  bool triangulate = !choose_parameter(get_parameter(np, internal_np::do_not_triangulate_faces), false);
 
-  //extend the bbox a bit to avoid border cases
-  double xd=(std::max)(1.,(bbox.xmax()-bbox.xmin())/100);
-  double yd=(std::max)(1.,(bbox.ymax()-bbox.ymin())/100);
-  double zd=(std::max)(1.,(bbox.zmax()-bbox.zmin())/100);
-  bbox=CGAL::Bbox_3(bbox.xmin()-xd, bbox.ymin()-yd, bbox.zmin()-zd,
-                    bbox.xmax()+xd, bbox.ymax()+yd, bbox.zmax()+zd);
-  TriangleMesh clipper;
-  Oriented_side os = internal::clip_to_bbox(plane, bbox, clipper, parameters::default_values());
-  switch(os)
+  auto vos = get(dynamic_vertex_property_t<Oriented_side>(), pm);
+  auto ecm = get(dynamic_edge_property_t<bool>(), pm, false);
+
+  if (triangulate && !is_triangle_mesh(pm))
+    triangulate = false;
+
+  refine_with_plane(pm, plane, parameters::vertex_oriented_side_map(vos)
+                                          .edge_is_marked_map(ecm)
+                                          .vertex_point_map(vpm)
+                                          .geom_traits(traits)
+                                          .do_not_triangulate_faces(!triangulate)
+                                          .throw_on_self_intersection(!allow_self_intersections &&
+                                                                      throw_on_self_intersection)
+                                          .concurrency_tag(Concurrency_tag()));
+
+
+  if (allow_self_intersections)
+    clip_volume=false;
+
+  if (clip_volume && !is_closed(pm)) clip_volume=false;
+  if (clip_volume && !use_compact_clipper) use_compact_clipper=true;
+
+
+  auto fcc = get(dynamic_face_property_t<std::size_t>(), pm);
+
+  std::size_t nbcc = connected_components(pm, fcc, CGAL::parameters::edge_is_constrained_map(ecm));
+
+  std::vector<bool> classified(nbcc, false);
+  std::vector<std::size_t> ccs_to_remove;
+
+  for (auto f : faces(pm))
   {
-    case ON_NEGATIVE_SIDE:
-      return true; // nothing to clip, the full mesh is on the negative side
-    case ON_POSITIVE_SIDE:
-      remove_all_elements(tm); // clear the mesh that is fully on the positive side
-      return true;
-    default:
-      break;
+    std::size_t ccid = get(fcc, f);
+    if (classified[ccid]) continue;
+    halfedge_descriptor hf = halfedge(f, pm);
+    for(halfedge_descriptor h : CGAL::halfedges_around_face(hf, pm))
+    {
+      CGAL::Oriented_side os = get(vos, target(h,pm));
```

Comment:
Kernel's Oriented_side rather than CGAL::'s ?

Reply:
we need to be certain in this context and won't get anything from Uncertain so it is correct.